### ISSUE TITLE
Support for older Android systems

### ISSUE THREAD

**Issue by gleb-skobinsky** [chrisbanes/haze#140]
Many thanks to you Chris and other contributors for this awesome library! It works like charm, unlike numerous fancy, but intangible solutions available out there. 
This is not an issue, but rather a feature request. How likely it is that this library may support RenderScript-based blur for the older Android versions (pre-Android 12)? Is a pull request to add such functionality something that you guys would consider?

**Comment by chrisbanes** [comment]
> How likely it is that this library may support RenderScript-based blur for the older Android versions (pre-Android 12)?

Not likely, unless someone contributes it.

> Is a pull request to add such functionality something that you guys would consider?

I'd definitely consider it! My only thought is that this would probably need to be opt-in.

**Comment by gleb-skobinsky** [comment]
Great, I will give it a try then sometime soon.

**Comment by desugar-64** [comment]
Just to add my two cents, I usually avoid using software (CPU) blurring on Android for tasks that need frequent updates. For static blur, where you blur an image or background once and leave it, it’s fine. But for real-time blur, where the background needs to update quickly with UI changes, software blur is just too resource-intensive and slow. It drains too much energy, memory, and CPU, and doesn’t perform well on less powerful devices.

It’s better to avoid CPU blurring on older APIs and use a simple scrim color instead. If it is unavoidable and you need to support blurring on APIs below 31 without sacrificing performance or energy consumption, the best approach is to use GPU-based blurring. It’s fast and power-efficient when implemented correctly, though it is more complex. I’m exploring this area, and the results are quite promising so far.

**Comment by chrisbanes** [comment]
+1 to @desugar-64. Any solution which relies on software rendering (i.e. drawing to a bitmap and using RenderScript) isn't going to be fast enough.

I'm interested to have a look at your GPU backed implementation @desugar-64. I assume that it's a OpenGL shader?

API 32 isn't just optimal because they have RenderEffects, but because we have access to `RenderNode`. That makes drawing content multiple times basically free. GraphicsLayer does have a back-compat version which uses a `View` underneath on older platforms, but I haven't used it enough if there's any performance trade-offs.

**Comment by desugar-64** [comment]
Hey, @chrisbanes. 

Yes, you’re correct. I’m using OpenGL ES for my implementation. I’ve written a rendering pipeline that performs post-processing on UI elements rendered into textures. Essentially, I’m using the GraphicsLayer to render UI into a surface-backed canvas, and then I use this texture for further post-processing.

I’ve open-sourced my project, and you can check it out [Imla](https://github.com/desugar-64/imla). 
I update it from time to time, so feel free to explore it. I hope you find it useful 🙂. 
Please note that I’m still actively working on it, so the code might not be perfectly clean yet. This is a hobby project of mine, so any feedback is welcome.

**Comment by desugar-64** [comment]
🎉

**Comment by desugar-64** [comment]
By the way, I have some additional ideas for improving rendering latency on API levels 29 and 30 (Android 10 and 11, respectively). Instead of copying data into a software bitmap from the output allocation, as we currently do, we could output the result directly into another surface created by a [HardwareBuffer](https://developer.android.com/reference/android/hardware/HardwareBuffer). This hardware buffer could then be [wrapped](https://developer.android.com/reference/kotlin/android/graphics/Bitmap?hl=en#wraphardwarebuffer) with a [hardware-configured bitmap](https://developer.android.com/reference/android/graphics/Bitmap.Config?hl=en#HARDWARE). 
Adopting this approach should eliminate an extra data copy behind RenderScript, further optimizing performance.

**Comment by chrisbanes** [comment]
Ah cool! I'm not entirely sure that `HardwareBuffer` and `Allocation` can work together, but it's definitely worth a try!

**Comment by desugar-64** [comment]
Yeah, it’s not directly bindable to a hardware buffer, but here’s some pseudo-code showing how it should work.

```kotlin
val outputAlloc: Allocation = ...
val ir = ImageReader.newInstance(width, height, PixelFormat.RGBA_8888, numImagesInQueue, HardwareBuffer.USAGE_GPU_SAMPLED_IMAGE).apply {
        setOnImageAvailableListener({ imageReader ->
            outputBitmap = Bitmap.wrapHardwareBuffer(imageReader.acquireNextImage().hardwareBuffer, ColorSpace.get(ColorSpace.Named.SRGB))
        }, null)
}
rsOutputAlloc.surface = ir.surface
```

I might also ask Romain if it’s actually worth it.

**Comment by desugar-64** [comment]
That lines up with what I thought.

![Image](https://github.com/user-attachments/assets/a1f90ff4-5965-4235-8cb6-daecb201acc9)

**Comment by Skaldebane** [comment]
Tried this on Android 10 (emulator), and it's working pretty well! The blur lags behind a bit, but that's expected as the docs clarify.

However there's one place where it looked quite weird, which was on resizing an item. I'm not sure what's happening here, but I'd guess that Haze is only rendering the blur for the visible size? And thus creating the blur for the expanded area also lags a bit?

Here's a video:

https://github.com/user-attachments/assets/437f8169-b9f8-4aef-ac90-e7203505f750

_(don't mind the freezing in the video, my system was under heavy load so the recorder lost some frames)_

Note that the blur effect on both the top bar and the background have the same source (which is the lazy list + fab), so that's definitely not the shadow.

Note: not posting this as a "bug" that needs fixing, but it was just interesting.
For now I'll just remove the blur from the top bar and make it semi-transparent (so it'll "feel" blurred because of the background).
